editor/gui/editor_scene_tabs.cpp

```diff
@@ -0,0 +1,375 @@
+	for (int i = 0; i < EditorNode::get_editor_data().get_edited_scene_count(); i++) {
+		Node *type_node = EditorNode::get_editor_data().get_edited_scene_root(i);
+		Ref<Texture2D> icon;
+		if (type_node) {
+			icon = EditorNode::get_singleton()->get_object_icon(type_node, "Node");
+		}
+
+		bool unsaved = EditorUndoRedoManager::get_singleton()->is_history_unsaved(EditorNode::get_editor_data().get_scene_history_id(i));
+		scene_tabs->add_tab(disambiguated_scene_names[i] + (unsaved ? "(*)" : ""), icon);
+
+		if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_GLOBAL_MENU)) {
+			DisplayServer::get_singleton()->global_menu_add_item("_dock", EditorNode::get_editor_data().get_scene_title(i) + (unsaved ? "(*)" : ""), callable_mp(this, &EditorSceneTabs::_global_menu_scene), Callable(), i);
+		}
+
+		if (show_rb && EditorNode::get_editor_data().get_scene_root_script(i).is_valid()) {
+			scene_tabs->set_tab_button_icon(i, script_icon);
+		}
+	}
+
+	if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_GLOBAL_MENU)) {
+		DisplayServer::get_singleton()->global_menu_add_separator("_dock");
+		DisplayServer::get_singleton()->global_menu_add_item("_dock", TTR("New Window"), callable_mp(this, &EditorSceneTabs::_global_menu_new_window));
+	}
+
+	if (scene_tabs->get_tab_count() > 0) {
+		scene_tabs->set_current_tab(EditorNode::get_editor_data().get_edited_scene());
+	}
+
+	const Size2 add_button_size = Size2(scene_tab_add->get_size().x, scene_tabs->get_size().y);
+	if (scene_tabs->get_offset_buttons_visible()) {
+		// Move the add button to a fixed position.
+		if (scene_tab_add->get_parent() == scene_tabs) {
+			scene_tabs->remove_child(scene_tab_add);
+			scene_tab_add_ph->add_child(scene_tab_add);
+			scene_tab_add->set_rect(Rect2(Point2(), add_button_size));
+		}
+	} else {
+		// Move the add button to be after the last tab.
+		if (scene_tab_add->get_parent() == scene_tab_add_ph) {
+			scene_tab_add_ph->remove_child(scene_tab_add);
+			scene_tabs->add_child(scene_tab_add);
+		}
+
+		if (scene_tabs->get_tab_count() == 0) {
+			scene_tab_add->set_rect(Rect2(Point2(), add_button_size));
+			return;
+		}
+
+		Rect2 last_tab = scene_tabs->get_tab_rect(scene_tabs->get_tab_count() - 1);
+		int hsep = scene_tabs->get_theme_constant(SNAME("h_separation"));
+		if (scene_tabs->is_layout_rtl()) {
+			scene_tab_add->set_rect(Rect2(Point2(last_tab.position.x - add_button_size.x - hsep, last_tab.position.y), add_button_size));
+		} else {
+			scene_tab_add->set_rect(Rect2(Point2(last_tab.position.x + last_tab.size.width + hsep, last_tab.position.y), add_button_size));
+		}
+	}
+
+	// Reconnect after everything is done.
+	scene_tabs->connect("tab_changed", callable_mp(this, &EditorSceneTabs::_scene_tab_changed));
+}
+
+void EditorSceneTabs::_tab_preview_done(const String &p_path, const Ref<Texture2D> &p_preview, const Ref<Texture2D> &p_small_preview, const Variant &p_udata) {
+	int p_tab = p_udata.operator signed int();
```

Comment:
```suggestion
	int p_tab = p_udata;
```
This can be simplified.